src/core/Akka.Streams/Implementation/Fusing/Ops.cs

```diff
@@ -3775,24 +3776,16 @@ private sealed class Logic : OutGraphStageLogic
             private readonly Action<T> _onSuccess;
             private readonly Action<Exception> _onFailure;
             private readonly Action _onComplete;
-            private readonly Action<Task<bool>> _handleContinuation;
+            private readonly CancellationTokenSource _completionCts;
 
-            public Logic(SourceShape<T> shape, IAsyncEnumerator<T> enumerator) : base(shape)
+            public Logic(SourceShape<T> shape, IAsyncEnumerable<T> enumerable) : base(shape)
             {
-                _enumerator = enumerator;
+                _completionCts = new CancellationTokenSource();
+                _enumerator = enumerable.GetAsyncEnumerator(_completionCts.Token);
```

Comment:
Interesting question, should this be in prestart vs the constructor?

Reply:
The only thing that would throw, I guess, would be if there was something wrong with the `IAsyncEnumerable` (i.e. `null`)?

In that case I guess it'd be better to have this run in `PreStart`